tests/core/BearTest.py

```diff
@@ -0,0 +1,262 @@
+from itertools import permutations
+from os.path import abspath, exists, isfile, join, getmtime
+import shutil
+from time import sleep
+import unittest
+
+from dependency_management.requirements.PipRequirement import PipRequirement
+
+import requests_mock
+
+from coalib.core.Bear import Bear
+from coalib.results.Result import Result
+from coalib.settings.Section import Section
+from coalib.settings.Setting import Setting
+
+
+class Bear1(Bear):
+    pass
+
+
+class Bear2(Bear):
+    pass
+
+
+class BearWithAnalysis(Bear):
+
+    def analyze(self, x: int, y: int, z: int=33):
+        """
+        Analyzes stuff.
+
+        :param x: First value.
+        :param y: Second value.
+        :param z: Third value.
+        """
+        yield x
+        yield y
+        yield z
+
+
+class BearWithMoreAnalysis(Bear):
+    BEAR_DEPS = {BearWithAnalysis}
+
+    def analyze(self, y: int, w: float):
+        """
+        Analyzes even more stuff.
+
+        :param y: Second value, but better.
+        :param w: Fourth value.
+        """
+        yield y
+        yield w
+
+
+class BearWithPrerequisites(Bear):
+    REQUIREMENTS = {PipRequirement('super-super-bad-package'),
+                    PipRequirement('even-even-worse-package')}
+
+
+class BearWithPrerequisitesOverride(Bear):
+    prerequisites_fulfilled = True
+
+    @classmethod
+    def check_prerequisites(cls):
+        return cls.prerequisites_fulfilled
+
+
+class BearTest(unittest.TestCase):
+
+    def tearDown(self):
+        defined_bears = [
+            value
+            for value in globals().values()
+            if isinstance(value, type) and issubclass(value, Bear)]
+
+        for bear in defined_bears:
+            if exists(bear.data_dir):
+                shutil.rmtree(bear.data_dir)
+
+    def test_invalid_types_at_instantiation(self):
+        with self.assertRaises(TypeError):
+            Bear(Section('test-section'), 2)
+
+        with self.assertRaises(TypeError):
+            Bear(None, {})
+
+    def test_analyze(self):
+        with self.assertRaises(NotImplementedError):
+            Bear(Section('test-section'), {}).analyze()
+
+    def test_generate_tasks(self):
+        with self.assertRaises(NotImplementedError):
+            Bear(Section('test-section'), {}).generate_tasks()
+
+    def test_add_dependency_results(self):
+        section = Section('test-section')
+        filedict = {}
+
+        uut = Bear(section, filedict)
+
+        self.assertEqual(uut.dependency_results,
+                         {})
+
+        bear1 = Bear1(section, filedict)
+        uut.add_dependency_results(bear1, [1, 2, 3])
+
+        self.assertEqual(uut.dependency_results,
+                         {Bear1: [1, 2, 3]})
+
+        bear2 = Bear2(section, filedict)
+        uut.add_dependency_results(bear2, [4, 5, 6])
+
+        self.assertEqual(uut.dependency_results,
+                         {Bear1: [1, 2, 3], Bear2: [4, 5, 6]})
+
+        uut.add_dependency_results(bear1, [7, 8, 9])
```

Comment:
The bear shouldn't yield multiple results normally. And if he does, it shouldn't be a problem anyway, then it seems to have a reason.

Actually I wanted to use sets before, but: In my current design I allow on purpose arbitrary types as results (coala enforces to use normal `Result` objects to get something meaningful). This can be convenient for dependency results, I have the feeling `HiddenResult`s contain too much info the dependent bear does not need. So why not just giving a tuple or a different type with the exact info what's needed, not less not more :)
Arbitrary types are not always hashable (consider dicts), also the order of the results can be important in a few cases (rare, but we support it anyway).